packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
+    super();
+    this.partialObjects = options.partialObjects ?? false;
+
+    if (Match.isMatcher(this.pattern)) {
+      throw new Error('ExactMatch cannot be nested with another matcher at the top level. Deeper nesting is allowed.');
+    }
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (Array.isArray(this.pattern)) {
+      return new ArrayMatch(this.pattern, { partial: false }).test(actual);
+    }
+
+    if (typeof this.pattern === 'object') {
+      return new ObjectMatch(this.pattern, { partial: this.partialObjects }).test(actual);
+    }
+
+    if (typeof this.pattern !== typeof actual) {
+      return [{ path: [], message: `Expected type ${typeof this.pattern} but received ${getType(actual)}` }];
+    }
+
+    if (this.pattern === ABSENT) {
+      throw new Error('ABSENT can only be used in an object matcher');
+    }
+
+    if (actual !== this.pattern) {
+      return [{ path: [], message: `Expected ${this.pattern} but received ${actual}` }];
+    }
+
+    return [];
+  }
+}
+
+/**
+ * Options when initializing the `ArrayMatch` class.
+ */
+export interface ArrayMatchOptions {
+  /**
+   * Whether the pattern should partially match with the target.
+   * The target array can contain more elements than expected by the pattern.
+   * Matching elements in the target must be present in the same order as in the pattern.
+   * @default true
+   */
+  readonly partial?: boolean;
+}
+
+/**
+ * Match class that matches arrays.
+ */
+export class ArrayMatch extends Match {
+  private readonly partial: boolean;
+
+  constructor(private readonly pattern: any[], options: ArrayMatchOptions = {}) {
+    super();
+    this.partial = options.partial ?? true;
+  }
+
+  public test(actual: any): MatchFailure[] {
+    if (!Array.isArray(actual)) {
+      return [{ path: [], message: `Expected type array but received ${getType(actual)}` }];
+    }
+    if (!this.partial && this.pattern.length !== actual.length) {
+      return [{ path: [], message: `Expected array of length ${this.pattern.length} but received ${actual.length}` }];
+    }
+
+    let patternIdx = 0;
+    let actualIdx = 0;
+
+    const failures: MatchFailure[] = [];
+    while (patternIdx < this.pattern.length && actualIdx < actual.length) {
+      let patternElement = this.pattern[patternIdx];
+      let matcher = Match.isMatcher(patternElement) ? patternElement : new LiteralMatch(patternElement);
+      const innerFailures = matcher.test(actual[actualIdx]);
+
+      if (!this.partial || innerFailures.length === 0) {
+        failures.push(...composeFailures(`[${actualIdx}]`, innerFailures));
+        patternIdx++;
+        actualIdx++;
+      } else {
+        actualIdx++;
+      }
+    }
+
+    for (; patternIdx < this.pattern.length; patternIdx++) {
+      const p = this.pattern[patternIdx];
+      const e = (Match.isMatcher(p) || typeof p === 'object') ? ' ' : ` [${p}] `;
+      failures.push({ path: [], message: `Missing element${e}at pattern index ${patternIdx}` });
```

Comment:
This touches on a big problem that the current matchers have: if we DIDN'T match, we can not report any useful information any more to the user.

Ideally, we'd print information on the CLOSEST match we found, by some distance metric that we need to invent. Maybe how deep it got into the tree?

Reply:
If we don't want to make the decision right here right now, we should leave the option open to add that later.

That means:

* Either report the entire matching tree back with all alternatives tried & all errors encountered, so that the reporting mechanism has all the information it could possibly use to construct the best error message
* Encapsulate the return value into some class type so that the data it tracks and stores is hidden and can be changed and enhanced at any point in time.

Reply:
Maybe flat count of errors is good enough as distance function...

Reply:
Yes. I struggled with this a little. I couldn't find a simple solution to this. Instead, I opted to use error count as a distance function. We do need to improve on this, but I was hoping to do it a little more incrementally/iteratively.

> we should leave the option open to add that later

Why do we need either of these options?
I was going to simply hide the distance judgement within the `Match` class and return the final result.

Reply:
> I was going to simply hide the distance judgement within the Match class and return the final result.

Obviously that, however you're incidentally also exposing the data you're using to make that decision, which means the data is going to be hard to change.